jsxImportSource is too eagerly parsed

# Bug Report



### 🔎 Search Terms

jsxImportSource

### 🕗 Version & Regression Information

- `4.3.0-dev.20210319`
- `4.2.3`

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about jsxImportSource

### ⏯ Playground Link

Not useful since the playground can't squire the types. I've created https://github.com/eps1lon/emotion-jsx-import-source-typescript instead

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/PQKgBApgzgNglgOwC4FoAmcoEMBGMJgBOEWAxksMWaoigFZQAeKUpA9gA4EopgCqCLAHcsxMGwBmYAAINGASQC2HNoSQBlNgFdCpAiGAAoUOFlMlKtZp16ZERWyRw2CSiXJgDhuMtVIwAN5gpFBQYAC+YBKEbIpgAETS9o7OrlTk8QDchoYQjJb+aBASWFow-hJaCOSpYADCsSoIEMgAFACUgYZgRBBIOghgrd09YAA8iBxaSCOjwaEAvAEhUAAGs3PBbDCqAFy9aNmbYKvhG8AAfCPt2WdAA)

### 💻 Code

```tsx
/* eslint-disable react/react-in-jsx-scope -- Unaware of @jsxImportSource */
/** @jsxImportSource @emotion/react */
import { css } from "@emotion/react";

export default function Component() {
  return (
    <input
      css={css`
        color: red;
      `}
    />
  );
}

```

### 🙁 Actual behavior

Compiler fails with `Cannot find module '*//jsx-runtime' or its corresponding type declarations.ts(2307)`

### 🙂 Expected behavior

Passes

### Context

I've identified the bug and can easily change it (by referring to `jsxImportSource` instead of `@jsxImportSource`) but the behavior is still unexpected and opaque. I suspect that altering the parser to not be as eager might be difficult. Instead an additional hint in the error message which line the compiler used to find the `@jsxImportSource` might be sufficient?

--- a/src/compiler/utilities.ts
+++ b/src/compiler/utilities.ts
@@ -6054,7 +6054,7 @@ namespace ts {
 
     export function getJSXImplicitImportBase(compilerOptions: CompilerOptions, file?: SourceFile): string | undefined {
         const jsxImportSourcePragmas = file?.pragmas.get("jsximportsource");
-        const jsxImportSourcePragma = isArray(jsxImportSourcePragmas) ? jsxImportSourcePragmas[0] : jsxImportSourcePragmas;
+        const jsxImportSourcePragma = isArray(jsxImportSourcePragmas) ? jsxImportSourcePragmas[jsxImportSourcePragmas.length - 1] : jsxImportSourcePragmas;
         return compilerOptions.jsx === JsxEmit.ReactJSX ||
             compilerOptions.jsx === JsxEmit.ReactJSXDev ||
             compilerOptions.jsxImportSource ||

--- a/tests/baselines/reference/jsxImportSourceNonPragmaComment.errors.txt
+++ b/tests/baselines/reference/jsxImportSourceNonPragmaComment.errors.txt
@@ -1,0 +1,26 @@
+tests/cases/compiler/jsxImportSourceNonPragmaComment.tsx(3,21): error TS2307: Cannot find module '@emotion/react' or its corresponding type declarations.
+tests/cases/compiler/jsxImportSourceNonPragmaComment.tsx(7,5): error TS2307: Cannot find module '@emotion/react/jsx-runtime' or its corresponding type declarations.
+
+
+==== tests/cases/compiler/jsxImportSourceNonPragmaComment.tsx (2 errors) ====
+    /* eslint-disable react/react-in-jsx-scope -- Unaware of @jsxImportSource */
+    /** @jsxImportSource @emotion/react */
+    import { css } from "@emotion/react";
+                        ~~~~~~~~~~~~~~~~
+!!! error TS2307: Cannot find module '@emotion/react' or its corresponding type declarations.
+    
+    export default function Component() {
+      return (
+        <input
+        ~~~~~~
+          css={css`
+    ~~~~~~~~~~~~~~~
+            color: red;
+    ~~~~~~~~~~~~~~~~~~~
+          `}
+    ~~~~~~~~
+        />
+    ~~~~~~
+!!! error TS2307: Cannot find module '@emotion/react/jsx-runtime' or its corresponding type declarations.
+      );
+    }

--- a/tests/baselines/reference/jsxImportSourceNonPragmaComment.js
+++ b/tests/baselines/reference/jsxImportSourceNonPragmaComment.js
@@ -1,0 +1,30 @@
+//// [jsxImportSourceNonPragmaComment.tsx]
+/* eslint-disable react/react-in-jsx-scope -- Unaware of @jsxImportSource */
+/** @jsxImportSource @emotion/react */
+import { css } from "@emotion/react";
+
+export default function Component() {
+  return (
+    <input
+      css={css`
+        color: red;
+      `}
+    />
+  );
+}
+
+//// [jsxImportSourceNonPragmaComment.jsx]
+"use strict";
+var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
+    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
+    return cooked;
+};
+exports.__esModule = true;
+/* eslint-disable react/react-in-jsx-scope -- Unaware of @jsxImportSource */
+/** @jsxImportSource @emotion/react */
+var react_1 = require("@emotion/react");
+function Component() {
+    return (<input css={(0, react_1.css)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n        color: red;\n      "], ["\n        color: red;\n      "])))}/>);
+}
+exports["default"] = Component;
+var templateObject_1;

--- a/tests/baselines/reference/jsxImportSourceNonPragmaComment.symbols
+++ b/tests/baselines/reference/jsxImportSourceNonPragmaComment.symbols
@@ -1,0 +1,20 @@
+=== tests/cases/compiler/jsxImportSourceNonPragmaComment.tsx ===
+/* eslint-disable react/react-in-jsx-scope -- Unaware of @jsxImportSource */
+/** @jsxImportSource @emotion/react */
+import { css } from "@emotion/react";
+>css : Symbol(css, Decl(jsxImportSourceNonPragmaComment.tsx, 2, 8))
+
+export default function Component() {
+>Component : Symbol(Component, Decl(jsxImportSourceNonPragmaComment.tsx, 2, 37))
+
+  return (
+    <input
+      css={css`
+>css : Symbol(css, Decl(jsxImportSourceNonPragmaComment.tsx, 6, 10))
+>css : Symbol(css, Decl(jsxImportSourceNonPragmaComment.tsx, 2, 8))
+
+        color: red;
+      `}
+    />
+  );
+}

--- a/tests/baselines/reference/jsxImportSourceNonPragmaComment.types
+++ b/tests/baselines/reference/jsxImportSourceNonPragmaComment.types
@@ -1,0 +1,27 @@
+=== tests/cases/compiler/jsxImportSourceNonPragmaComment.tsx ===
+/* eslint-disable react/react-in-jsx-scope -- Unaware of @jsxImportSource */
+/** @jsxImportSource @emotion/react */
+import { css } from "@emotion/react";
+>css : any
+
+export default function Component() {
+>Component : () => any
+
+  return (
+>(    <input      css={css`        color: red;      `}    />  ) : any
+
+    <input
+><input      css={css`        color: red;      `}    /> : any
+>input : any
+
+      css={css`
+>css : any
+>css`        color: red;      ` : any
+>css : any
+>`        color: red;      ` : "\n        color: red;\n      "
+
+        color: red;
+      `}
+    />
+  );
+}

--- a/tests/cases/compiler/jsxImportSourceNonPragmaComment.tsx
+++ b/tests/cases/compiler/jsxImportSourceNonPragmaComment.tsx
@@ -1,0 +1,14 @@
+// @jsx: preserve
+/* eslint-disable react/react-in-jsx-scope -- Unaware of @jsxImportSource */
+/** @jsxImportSource @emotion/react */
+import { css } from "@emotion/react";
+
+export default function Component() {
+  return (
+    <input
+      css={css`
+        color: red;
+      `}
+    />
+  );
+}

